File: src/test/java/night/kotlin/seminar/javaexample/lambda/LambdaExampleTest.java
package night.kotlin.seminar.javaexample.lambda;

import static org.assertj.core.api.Java6Assertions.*;

import java.util.function.DoubleToIntFunction;
import java.util.function.DoubleToLongFunction;
import java.util.function.Function;
import java.util.function.LongFunction;

import org.junit.jupiter.api.Test;

class LambdaExampleTest {

	@Test
	public void testHighOrderFunction() {
		LambdaExample exam = new LambdaExample();

		Function<Long, Long> addFunction = exam.createAddFrom(5L);
		assertThat(addFunction.apply(10L)).isEqualTo(15L);
		assertThat(addFunction.apply(20L)).isEqualTo(25L);
	}

	@Test
	public void 자바는_함수대신_Function과_같은_타입으로_함수를_담을수_있다() {
		LongFunction function = (num) -> num + 4L;
		LongFunction function과동일 = new LongFunction() {
			@Override
			public Object apply(long value) {
				return value + 4L;
			}
		};

		System.out.println(function.getClass());
		System.out.println(function과동일.getClass());

		assertThat(function.apply(1L)).isEqualTo(5L);
		assertThat(function과동일.apply(1L)).isEqualTo(5L);
	}
}

File: src/test/kotlin/night/kotlin/seminar/kotlin/lambda/LambdaExampleTest.kt
package night.kotlin.seminar.kotlin.lambda

import org.assertj.core.api.Java6Assertions.assertThat
import org.junit.jupiter.api.Test

internal class LambdaExampleTest {
    @Test
    fun `코틀린은 펑션을 리턴할 할 수 있다`() {
        assertThat(LambdaExample.add(1).invoke(1L)).isEqualTo(2L)
        assertThat(LambdaExample.add1(1).invoke(1L)).isEqualTo(2L)
        assertThat(LambdaExample.add2(1).invoke(1L)).isEqualTo(2L)

        // invoke 없이 직접 호출 가능
        assertThat(LambdaExample.add(1)(2)).isEqualTo(3L)

        // currying 형태의 함수 형태로도 지원
        assertThat(LambdaExample.add3 { 1 }(2)(3L)).isEqualTo(6L)
        // 위 라인을 풀어쓴 표현
        assertThat(LambdaExample.add3({ return@add3 1L })(2)(3)).isEqualTo(1 + 2 + 3L)
        // currying 형태의 변수 형태로도 지원
        assertThat(LambdaExample.add4 { 1 }(2)(3)).isEqualTo(1 + 2 + 3L)

        val addOne = LambdaExample.add4 { 1 }
        println(addOne)
        val addOnePlusFive = addOne(5)
        println(addOnePlusFive)
        assertThat(addOnePlusFive(7)).isEqualTo(1 + 5 + 7L)
        assertThat(addOne(8)(9)).isEqualTo(1 + 8 + 9L)
    }

    @Test
    fun `변수에 함수를 직접 담을 수 있고 함수 타입을 출력해보면 Function이 아닌 함수 타입이 출력된다`() {
        val func = { num: Long -> num + 4 }

        print(func)

        assertThat(func(1))
        assertThat(func.invoke(1))
    }

}